### DISCUSSION TITLE
Better naming for SimplePolygon and Polygons

### DISCUSSION THREAD

**Discussion by geoffder** [elalish/manifold#390] [Ideas]
I was thinking that it might be good to rename them to `Contour` and `Contours` respectively. That way they would line up a bit better with the way that they are talked about in the docs, and also convey more clearly that `Polygons` does not necessarily have any structure or relationship between the contours within (e.g. they can be intersecting and are not ordered into complex polygons in any way).

Does that make sense?

**Comment by elalish** [comment]
Yeah, that's reasonable, but it still feels like 6 of one, half-dozen of the other. I'd like to avoid breaking changes where possible, so I'd rather avoid a rename here.

Actually, this made me realize something - you swapped `Extrude` and `Revolve` to take `CrossSection` instead of `Polygons`. For the same reason, I think I'd like to revert that. It's also nice because it allows `Manifold` and `CrossSection` to be independent libraries without dependencies on each other. I try to keep the pieces separated that way to the greatest extent reasonable. Same reason `SDF` doesn't create a `Manifold` directly. 

How does that sound?

**Comment by elalish** [comment]
Well, I suppose `Cylinder` and such are still using `Circle`, so maybe not that independent, but meh.

**Comment by geoffder** [comment]
Ah, I figured the change to CrossSection was desirable (I think we did talk about it before I did it) since it enforces the inputs to be intersection free (something that is totally up the user otherwise). Also it is better ergonomics to be able to give CrossSections straight to Extrude and Revolve, rather than always having to use ToPolygons.

**Comment by geoffder** [comment]
In general I'm on the side of maximizing the ergonomics of the API over keeping things seperate in case someone wanted to spare themselves depending on the sublibraries (which are quite small compared to Manifold itself).

Tangentially related, I'm personally looking forward to when module support improves, I might like to take a crack at reorganizing with then and see how much faster it makes the build.

**Comment by geoffder** [comment]
Sorry to spam, but another thing came to mind that I'll add before I sleep and forget: RE breaking changes with Polygons, we have already recently changed the type by dropping the indexed struct. 

***If*** we come to an agreement about names that would be better going forward, at least it isn't something that is too stable. Plus they are *just* aliases as is, so the old names could be easily deprecated with the new existing alongside for some time.

**Comment by elalish** [comment]
You make good points. I'm thinking it over; I'll get back to you.

**Comment by elalish** [comment]
Okay, I agree about keeping Extrude and Revolve taking CrossSections - you're right that forcing removal of self-intersections is important for the ergonomics. 

Yes, I did just make a breaking change to Polygon, but I still feel like changing their names now mostly amounts to needless churn - Contours doesn't evoke much difference in meaning to me. I think we'd be better off including clear descriptive docs for those aliases instead. 

Thanks for mentioning C++20 modules - I ended up reading up on them. Agreed, that sounds great once the compilers are ready for us (and it sounds like they're pretty close now). That would be ideal, because definitely part of my library design here was focused on keeping the compile time down to something reasonable.

**Reply by geoffder** [reply]
Speaking of build times, I've opened an issue on Clipper2 to ask whether they'd be open to making the build of the `Clipper2Z` library (which we don't use) optional. It doesn't add too much compared to our total build time anyway (especially if CUDA is on), but I figure it'd be nice to shave off a bit there if we can without much fuss.

**Reply by pca006132** [reply]
I tried using pch but got some weird errors. I think the major slowness probably come from thrust, which is a template library. I don't think modules/pch does much to reduce compilation time for templates unfortunately.

**Reply by geoffder** [reply]
🤔 Now that you mention it, I wonder if switching from thrust to std parallel stuff like you guys have discussed recently would help with the compile times.

**Reply by elalish** [reply]
Honestly, at least for nvcc, I think they're basically using the same code for both so I doubt it makes much difference.

**Reply by geoffder** [reply]
Supposedly clang can compile CUDA, and I saw some comments saying the build was faster (though I'm not sure about performance of the code). I've actually never changed my make config to use it over gcc so now I'm curious to try. I seem to remember being recommended it for improved template errors.

**Reply by geoffder** [reply]
> Speaking of build times, I've opened an issue on Clipper2 to ask whether they'd be open to making the build of the `Clipper2Z` library (which we don't use) optional.

Well, that was a quick turnaround! https://github.com/AngusJohnson/Clipper2/pull/463
I'll have to update our Clipper2 commit and set the new flag then.
